### PR TITLE
Bump to latest rocket-chip master

### PR DESCRIPTION
**Pull request by pentin-as**
tested with rocket-chip a48dd575b23aa4b82368f19beaaf19e2d4bf1e3e
It is also required in the boom-template in the Makefraq file to write "runMain" instead of "run-main".

### CODE REVIEW

**Comment by ccelio** [comment]
I'm having difficulty reproducing a success with this patch.

I updated rocket-chip to (https://github.com/freechipsproject/rocket-chip/commit/a48dd575b23aa4b82368f19beaaf19e2d4bf1e3e), and changed the Makefrag from `run-main` to `runMain`. And I rebuilt the tools.

Are there other changes that need to be made?

```
g++  -I.  -MMD -I/home/eng/celio/work/merge-boom-template/verisim/verilator/install/share/verilator/include -I/home/eng/celio/work/merge-boom-template/verisim/verilator/install/share/verilator/include/vltstd -DVL_PRINTF=printf -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -Wno-char-subscripts -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable     -O1 -std=c++11 -I/home/eng/celio/install/riscv/09a1ff/include -D__STDC_FORMAT_MACROS -DVERILATOR -DTEST_HARNESS=VTestHarness -include /home/eng/celio/work/merge-boom-template/rocket-chip/csrc/verilator.h -I/home/eng/celio/work/merge-boom-template/verisim/generated-src -include /home/eng/celio/work/merge-boom-template/verisim/generated-src/boom.system.TestHarness.BoomConfig/VTestHarness.h   -c -o VTestHarness__7.o VTestHarness__7.cpp
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc: In function ‘void usage(const char*)’:
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc:85:14: error: expected ‘)’ before ‘PLUSARG_USAGE_OPTIONS’
   fputs("\n" PLUSARG_USAGE_OPTIONS, stdout);
              ^
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc:86:14: error: expected ‘)’ before ‘HTIF_USAGE_OPTIONS’
   fputs("\n" HTIF_USAGE_OPTIONS, stdout);
              ^
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc: In function ‘int main(int, char**)’:
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc:130:7: error: ‘HTIF_LONG_OPTIONS’ was not declared in this scope
       HTIF_LONG_OPTIONS
       ^
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc:184:36: error: ‘verilog_plusargs’ was not declared in this scope
           const char ** plusarg = &verilog_plusargs[0];
                                    ^
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc:221:18: error: ‘HTIF_LONG_OPTIONS_OPTIND’ was not declared in this scope
         if (c >= HTIF_LONG_OPTIONS_OPTIND) {
                  ^
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc:262:39: error: no matching function for call to ‘dtm_t::dtm_t(int&, char**&)’
   dtm = new dtm_t(htif_argc, htif_argv);
                                       ^
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc:262:39: note: candidate is:
In file included from /home/eng/celio/work/merge-boom-template/rocket-chip/csrc/emulator.cc:8:0:
/home/eng/celio/install/riscv/09a1ff/include/fesvr/dtm.h:17:3: note: dtm_t::dtm_t(const std::vector<std::basic_string<char> >&)
   dtm_t(const std::vector<std::string>& args);
   ^
/home/eng/celio/install/riscv/09a1ff/include/fesvr/dtm.h:17:3: note:   candidate expects 1 argument, 2 provided
make[1]: *** [emulator.o] Error 1
make[1]: *** Waiting for unfinished jobs....
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/SimDTM.cc: In function ‘int debug_tick(unsigned char*, unsigned char, int*, int*, int*, unsigned char, unsigned char*, int, int)’:
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/SimDTM.cc:26:43: error: no matching function for call to ‘dtm_t::dtm_t(PLI_INT32&, PLI_BYTE8**&)’
       dtm = new dtm_t(info.argc, info.argv);
                                           ^
/home/eng/celio/work/merge-boom-template/rocket-chip/csrc/SimDTM.cc:26:43: note: candidate is:
In file included from /home/eng/celio/work/merge-boom-template/rocket-chip/csrc/SimDTM.cc:3:0:
/home/eng/celio/install/riscv/09a1ff/include/fesvr/dtm.h:17:3: note: dtm_t::dtm_t(const std::vector<std::basic_string<char> >&)
   dtm_t(const std::vector<std::string>& args);
   ^
/home/eng/celio/install/riscv/09a1ff/include/fesvr/dtm.h:17:3: note:   candidate expects 1 argument, 2 provided
make[1]: *** [SimDTM.o] Error 1
make[1]: Leaving directory `/home/eng/celio/work/merge-boom-template/verisim/generated-src/boom.system.TestHarness.BoomConfig'
make: *** [/home/eng/celio/work/merge-boom-template/verisim/simulator-boom.system-BoomConfig] Error 2
```

**Comment by pentin-as** [comment]
I am very sorry. I tested on my test environment (without verilator).

1. In boom-template/verisim/Makefrag-verilator in line 35, you also need to add "include ...plusArgs":
  -O3 -CFLAGS "$(CXXFLAGS) -DVERILATOR -DTEST_HARNESS=V$(MODEL) -include $(base_dir)/rocket-chip/csrc/verilator.h **-include $(build_dir)/$(PROJECT).$(MODEL).$(CONFIG).plusArgs**"

2. In boom-template/Makefrag-verilator add the compilation SimJTAG and remote_bitbang:
sim_csrcs = \\
	$(base_dir)/rocket-chip/csrc/SimDTM.cc \\
 	**$(base_dir)/rocket-chip/csrc/SimJTAG.cc \\
	$(base_dir)/rocket-chip/csrc/remote_bitbang.cc \\**
...

**Comment by seldridge** [comment]
Sorry... I should have jumped in here earlier as this is all related to my idiotic effort to improve the visibility and parsing of options.

I think you're past it, but all the htif errors are due to a change from kludgily parsing `std::string` to `int argc, char ** argv`, i.e., `getopt`. 

Related to the `PlusArg`, any Rocket `PlusArg` that is defined is supposed to show up in the build directory in one of those *.plusArgs files. That is supposed to be pulled in by the updates to `emulator/Make*` to define a needed def `PLUSARG_USAGE_OPTIONS` so that `emulator.cc` knows what is a valid option and can generate pretty help text for people like me who can't remember what the commands are.

I'm trying to reproduce based on some information that @ccelio provided.

**Comment by ccelio** [comment]
Thanks @seldridge I'm 99% sure that my riscv-tools install went awry and pushed the wrong riscv-fesvr/htif.h to my $RISCV directory.

**Comment by ccelio** [comment]
And thanks for writing good comments in the csrc/emulator.cc so I knew where to look for missing macros.

**Comment by seldridge** [comment]
Following the guidelines from your email... Here's what I did (and it worked on my end... ugh):
```bash
git clone https://github.com/freechipsproject/rocket-chip.git rocket-chip-a48dd5
cd rocket-chip-a48dd5/
git checkout a48dd575b23aa4b82368f19beaaf19e2d4bf1e3e
git submodule update --init
export RISCV=/opt/riscv/stow/riscv-tools-2017-12-26-9868299
export PATH=$RISCV/bin:$PATH
make
./emulator-freechips.rocketchip.system-DefaultConfig
```

For the `PlusArg` part, what I'm seeing is that your `g++` call above isn't including the generated header. Expanding out your call:
```bash
g++  \
    -I. \
    -MMD \
    -I/home/eng/celio/work/merge-boom-template/verisim/verilator/install/share/verilator/include \
    -I/home/eng/celio/work/merge-boom-template/verisim/verilator/install/share/verilator/include/vltstd \
    -DVL_PRINTF=printf \
    -DVM_COVERAGE=0 \
    -DVM_SC=0 \
    -DVM_TRACE=0 \
    -Wno-char-subscripts \
    -Wno-sign-compare \
    -Wno-uninitialized \
    -Wno-unused-but-set-variable \
    -Wno-unused-parameter \
    -Wno-unused-variable \
    -O1 \
    -std=c++11 \
    -I/home/eng/celio/install/riscv/09a1ff/include \
    -D__STDC_FORMAT_MACROS \
    -DVERILATOR \
    -DTEST_HARNESS=VTestHarness \
    -include /home/eng/celio/work/merge-boom-template/rocket-chip/csrc/verilator.h \
    -I/home/eng/celio/work/merge-boom-template/verisim/generated-src \
    -include /home/eng/celio/work/merge-boom-template/verisim/generated-src/boom.system.TestHarness.BoomConfig/VTestHarness.h \
    -c \
    -o VTestHarness__7.o VTestHarness__7.cpp
```

And my call from that run just now:
```bash
g++  \
    -I.  \
    -MMD \
    -I/home/se/tmp/rocket-chip-a48dd5/emulator/verilator/install/share/verilator/include \
    -I/home/se/tmp/rocket-chip-a48dd5/emulator/verilator/install/share/verilator/include/vltstd \
    -DVL_PRINTF=printf \
    -DVM_COVERAGE=0 \
    -DVM_SC=0 \
    -DVM_TRACE=0 \
    -Wno-char-subscripts \
    -Wno-sign-compare \
    -Wno-uninitialized \
    -Wno-unused-but-set-variable \
    -Wno-unused-parameter \
    -Wno-unused-variable     \
    -O1 \
    -std=c++11 \
    -I/opt/riscv/stow/riscv-tools-2017-12-26-9868299/include \
    -I/home/se/tmp/rocket-chip-a48dd5/csrc \
    -DVERILATOR \
    -DTEST_HARNESS=VTestHarness \
    -include /home/se/tmp/rocket-chip-a48dd5/csrc/verilator.h \
    -include /home/se/tmp/rocket-chip-a48dd5/emulator/generated-src/freechips.rocketchip.system.DefaultConfig.plusArgs \
    -I/home/se/tmp/rocket-chip-a48dd5/emulator/generated-src \
    -include /home/se/tmp/rocket-chip-a48dd5/emulator/generated-src/freechips.rocketchip.system.DefaultConfig/VTestHarness.h   \
    -c -o VTestHarness__7.o VTestHarness__7.cpp
```

With the difference being `-include /home/se/tmp/rocket-chip-a48dd5/emulator/generated-src/freechips.rocketchip.system.DefaultConfig.plusArgs`. This is supposed to be added by [`Makefrag-verilog` here](https://github.com/freechipsproject/rocket-chip/blob/a48dd575b23aa4b82368f19beaaf19e2d4bf1e3e/emulator/Makefrag-verilator#L63), but it's not?

This makes me think that there's some version mismatch with that specific call above.

- `PLUSARG_USAGE_OPTIONS` should come from `emulator/generated-src/freechips.rocketchip.system.DefaultConfig.plusArgs`
- `HTIF_USAGE_OPTIONS` should come from `$RISCV/include/fesvr/htif.h`:
```
> grep -r HTIF_USAGE_OPTIONS $RISCV/include
/opt/riscv/stow/riscv-tools-2017-12-26-9868299/include/fesvr/htif.h:#define HTIF_USAGE_OPTIONS "HOST OPTIONS\n\
```

__tl;dr: This all makes me think that you're pointing at the wrong `htif.h` and have a bad `Makefrag-verilog`.__

**Comment by ccelio** [comment]
Thanks, I'm re-rebuilding my riscv-tools, and I'm carefully monitoring the installed htif.h. I'll let you know if I hit any more snags. Thanks for sanity checking this for me.

**Comment by ccelio** [comment]
Thanks everybody.
